sycl/plugins/cuda/pi_cuda.cpp

```diff
@@ -3396,17 +3396,32 @@ pi_result cuda_piEnqueueMemUnmap(pi_queue command_queue, pi_mem memobj,
 pi_result cuda_piextUSMHostAlloc(void **result_ptr, pi_context context,
                                  pi_usm_mem_properties *properties, size_t size,
                                  pi_uint32 alignment) {
+  static constexpr pi_uint32 max_alignment =
+      0x200; // from empirical testing with CUDA 10.2 on a Tesla K40
+
+  // enforce a valid pointer to the allocated memory
   assert(result_ptr != nullptr);
-  assert(context != nullptr);
-  assert(properties == nullptr);
+  // check the the context is valid
+  if (context == nullptr)
+    return PI_INVALID_CONTEXT;
+  // check that the property list is empty
+  if (properties != nullptr)
+    return PI_INVALID_PROPERTY;
+  // check that the alignment is not larger than max_alignment, and is either 0
+  // or a power of 2
+  if (alignment > max_alignment || __builtin_popcount(alignment) > 1)
```

Comment:
Is __builtin_popcount supported by GCC, CLANG, VC++ compilers? If no, can you implement it using regular c++?

Reply:
isn't this part always compiled by clang itself ?

Reply:
anyway, according to godbolt, it works on
  - gcc on x86, arm, risc-v
  - clang on x86, arm
  - msvc with a workaround:
    ```c++
    #ifdef _MSC_VER
    #include <intrin.h>
    #define __builtin_popcount __popcnt
    #endif
    ```

If this code really needs to be compiled by other compilers, I can add the workaround at the top of the file.

Reply:
Otherwise `alignment & (alignment - 1)` is just plain C while waiting for C++20 bit functions...

Reply:
Well, thanks.

That compiles to simpler code both with and without `popcnt` support:
![image](https://user-images.githubusercontent.com/4069793/80276502-8ed44780-86e9-11ea-955d-00a0875554d7.png)

Reply:
\o/ Thanks for having made the "benchmark"... :-)
Quite interesting.